File: app/src/test/java/com/example/popview/ComentarioViewModelTest.java
package com.example.popview;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;

import com.example.popview.data.Comentario
import com.example.popview.viewmodel.ComentarioViewModel;

import org.junit.Assert.*
import org.junit.Before
import org.junit.Test

class ComentarioViewModelTest {

    private lateinit var viewModel: ComentarioViewModel

    @Before
    fun setUp() {
        viewModel = ComentarioViewModel()
        viewModel.establecerUsuarioActual(1) // Usuario ficticio
    }

    @Test
    fun `validar texto vacío da error`() {
        viewModel.actualizarTextoComentario("")
        val estado = viewModel.estadoComentario.value

        assertEquals("El comentario no puede estar vacío", estado?.errorTexto)
        assertFalse(estado?.esValido ?: true)
    }

    @Test
    fun `validar puntuación inválida da error`() {
        viewModel.actualizarPuntuacion(5f)
        val estado = viewModel.estadoComentario.value

        assertEquals("La puntuación debe entre 0 y 4", estado?.errorPuntuacion)
        assertFalse(estado?.esValido ?: true)
    }

    @Test
    fun `enviar comentario válido lo genera correctamente`() {
        var comentarioEnviado: Comentario? = null

        viewModel.actualizarTextoComentario("Buena peli")
        viewModel.actualizarPuntuacion(3.5f)

        viewModel.enviarComentario(idTitulo = 1) {
            comentarioEnviado = it
        }

        assertNotNull(comentarioEnviado)
        assertEquals("Buena peli", comentarioEnviado?.comentaris)
        assertEquals(3.5f, comentarioEnviado?.rating)
        assertEquals(1, comentarioEnviado?.usuari_id)
        assertEquals(0, comentarioEnviado?.id) // Porque se autogenera después
    }

    @Test
    fun `editar comentario conserva el id y actualiza campos`() {
        var comentarioEditado: Comentario? = null

        viewModel.prepararEdicion(idComentario = 99, texto = "Antiguo", puntuacion = 1f)
        viewModel.actualizarTextoComentario("Actualizado")
        viewModel.actualizarPuntuacion(4f)

        viewModel.enviarComentario(idTitulo = 1) {
            comentarioEditado = it
        }

        assertNotNull(comentarioEditado)
        assertEquals(99, comentarioEditado?.id)
        assertEquals("Actualizado", comentarioEditado?.comentaris)
        assertEquals(4f, comentarioEditado?.rating)
    }

    @Test
    fun `agregar comentario se añade a la lista`() {
        val nuevo = Comentario(id = 1, usuari_id = 1, comentaris = "Test", rating = 3f)
        viewModel.agregarComentarioALista(nuevo)

        assertTrue(viewModel.comentarios.value!!.contains(nuevo))
    }

    @Test
    fun `editar comentario en lista lo reemplaza correctamente`() {
        val original = Comentario(id = 10, usuari_id = 1, comentaris = "Viejo", rating = 1f)
        val editado = Comentario(id = 10, usuari_id = 1, comentaris = "Nuevo", rating = 4f)

        viewModel.agregarComentarioALista(original)
        viewModel.editarComentarioEnLista(editado)

        val comentario = viewModel.comentarios.value!!.find { it.id == 10 }
        assertEquals("Nuevo", comentario?.comentaris)
        assertEquals(4f, comentario?.rating)
    }

    @Test
    fun `eliminar comentario lo quita solo a él`() {
        val comentario1 = Comentario(id = 1, usuari_id = 1, comentaris = "Primero", rating = 2f)
        val comentario2 = Comentario(id = 2, usuari_id = 1, comentaris = "Segundo", rating = 3f)

        viewModel.agregarComentarioALista(comentario1)
        viewModel.agregarComentarioALista(comentario2)

        viewModel.eliminarComentario(1)

        val lista = viewModel.comentarios.value!!
                assertFalse(lista.any { it.id == 1 })
        assertTrue(lista.any { it.id == 2 })
    }
}

File: app/src/test/java/com/example/popview/ComentarioViewModelTest.kt
package com.example.popview

import com.example.popview.viewmodel.ComentarioViewModel
import com.example.popview.data.Comentario
import junit.framework.TestCase.assertEquals
import junit.framework.TestCase.assertFalse
import junit.framework.TestCase.assertTrue
import org.junit.Before
import org.junit.Test

class ComentarioViewModelTest {

    private lateinit var viewModel: ComentarioViewModel

    @Before
    fun setUp() {
        viewModel = ComentarioViewModel()
    }

    @Test
    fun `comentario con campos vacíos o inválidos retorna errores de validación`() {
        val comentarioInvalido = Comentario(
            usuari_id = 0,
            comentaris = "",
            rating = -1f
        )

        val estado = viewModel.validarComentario(comentarioInvalido)

        assertFalse(estado.esValido)
        assertEquals("El comentario no puede estar vacío", estado.errorComentario)
        assertEquals("El ID de usuario debe ser mayor que 0", estado.errorUsuarioId)
        assertEquals("La puntuación debe estar entre 0 y 4", estado.errorRating)
    }

    @Test
    fun `agregar comentario válido lo añade a la lista`() {
        val comentarioValido = Comentario(
            usuari_id = 1,
            comentaris = "Este es un comentario de prueba",
            rating = 4.5f
        )

        val resultado = viewModel.agregarComentario(comentarioValido)
        assertTrue(resultado)
        assertEquals(1, viewModel.obtenerTodos().size)
    }

    @Test
    fun `modificar comentario válido actualiza la lista`() {
        val comentarioValido = Comentario(
            id = 1,
            usuari_id = 1,
            comentaris = "Comentario original",
            rating = 3.0f
        )
        viewModel.agregarComentario(comentarioValido)

        val resultado = viewModel.modificarComentario(1, "Comentario modificado", 4.0f)
        assertTrue(resultado)
        assertEquals("Comentario modificado", viewModel.obtenerTodos()[0].comentaris)
        assertEquals(4.0f, viewModel.obtenerTodos()[0].rating)
    }

    @Test
    fun `eliminar comentario existente lo remueve de la lista`() {
        val comentarioValido = Comentario(
            id = 1,
            usuari_id = 1,
            comentaris = "Comentario a eliminar",
            rating = 3.5f
        )
        viewModel.agregarComentario(comentarioValido)

        val resultado = viewModel.eliminarComentario(1)
        assertTrue(resultado)
        assertTrue(viewModel.obtenerTodos().isEmpty())
    }

    @Test
    fun `inicialmente la lista de comentarios está vacía`() {
        assertTrue(viewModel.obtenerTodos().isEmpty())
    }

    @Test
    fun `comentario válido no tiene errores de validación`() {
        val comentarioValido = Comentario(
            usuari_id = 1,
            comentaris = "Comentario válido",
            rating = 4.0f
        )

        val estado = viewModel.validarComentario(comentarioValido)

        assertTrue(estado.esValido)
        assertEquals(null, estado.errorComentario)
        assertEquals(null, estado.errorUsuarioId)
        assertEquals(null, estado.errorRating)
    }

    @Test
    fun `comentario con comentaris de longitud máxima permitida es válido`() {
        val comentarioValido = Comentario(
            usuari_id = 1,
            comentaris = "a".repeat(500), // Asumiendo que 500 es la longitud máxima permitida
            rating = 3.0f
        )

        val estado = viewModel.validarComentario(comentarioValido)

        assertTrue(estado.esValido)
        assertEquals(null, estado.errorComentario)
    }

    @Test
    fun `agregar comentario con comentaris vacío retorna error`() {
        val comentarioInvalido = Comentario(
            usuari_id = 1,
            comentaris = "",
            rating = 3.0f
        )

        val estado = viewModel.validarComentario(comentarioInvalido)

        assertFalse(estado.esValido)
        assertEquals("El comentario no puede estar vacío", estado.errorComentario)
    }

    @Test
    fun `agregar múltiples comentarios válidos incrementa el tamaño de la lista`() {
        val comentario1 = Comentario(
            usuari_id = 1,
            comentaris = "Primer comentario",
            rating = 4.0f
        )
        val comentario2 = Comentario(
            usuari_id = 2,
            comentaris = "Segundo comentario",
            rating = 3.5f
        )

        viewModel.agregarComentario(comentario1)
        viewModel.agregarComentario(comentario2)

        assertEquals(2, viewModel.obtenerTodos().size)
    }
}
